src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs

```diff
@@ -23,14 +25,15 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
     /// in that they watch the running document table, tracking open/close events, and also file
     /// change events while the file is closed.
     /// </summary>
-    internal abstract partial class DocumentProvider
+    internal abstract partial class DocumentProvider : ForegroundThreadAffinitizedObject
     {
         protected readonly IVsRunningDocumentTable4 RunningDocumentTable;
         protected readonly bool IsRoslynPackageInstalled;
         protected readonly IVsEditorAdaptersFactoryService EditorAdaptersFactoryService;
         protected readonly IContentTypeRegistryService ContentTypeRegistryService;
 
         private readonly uint _runningDocumentTableEventCookie;
+        private readonly object _gate = new object();
```

Comment:
Can you comment which fields this covers? I like how you did it for the other types.